subcommand/consul-sidecar/command.go

```diff
@@ -80,57 +103,166 @@ func (c *Command) Run(args []string) int {
 		c.UI.Error(err.Error())
 		return 1
 	}
+	c.logger = logger
 
 	// Log initial configuration
-	logger.Info("Command configuration", "service-config", c.flagServiceConfig,
+	c.logger.Info("Command configuration", "service-config", c.flagServiceConfig,
 		"consul-binary", c.flagConsulBinary,
 		"sync-period", c.flagSyncPeriod,
-		"log-level", c.flagLogLevel)
+		"log-level", c.flagLogLevel,
+		"enable-metrics-merging", c.flagEnableMetricsMerging,
+		"merged-metrics-port", c.flagMergedMetricsPort,
+		"service-metrics-port", c.flagServiceMetricsPort,
+		"service-metrics-path", c.flagServiceMetricsPath,
+	)
 
 	c.consulCommand = []string{"services", "register"}
 	c.consulCommand = append(c.consulCommand, c.parseConsulFlags()...)
 	c.consulCommand = append(c.consulCommand, c.flagServiceConfig)
 
-	// ctx that we pass in to the main work loop, signal handling is handled in another thread
+	// signalCtx that we pass in to the main work loop, signal handling is handled in another thread
 	// due to the length of time it can take for the cmd to complete causing synchronization issues
 	// on shutdown. Also passing a context in so that it can interrupt the cmd and exit cleanly.
-	ctx, cancelFunc := context.WithCancel(context.Background())
+	signalCtx, cancelFunc := context.WithCancel(context.Background())
 	go func() {
 		select {
 		case sig := <-c.sigCh:
-			logger.Info(fmt.Sprintf("%s received, shutting down", sig))
+			c.logger.Info(fmt.Sprintf("%s received, shutting down", sig))
 			cancelFunc()
 			return
 		}
 	}()
-	// The main work loop. We continually re-register our service every
-	// syncPeriod. Consul is smart enough to know when the service hasn't changed
-	// and so won't update any indices. This means we won't be causing a lot
-	// of traffic within the cluster. We tolerate Consul Clients going down
-	// and will simply re-register once it's back up.
+
+	// If metrics merging is enabled, run a merged metrics server in a goroutine
+	// that serves Envoy sidecar metrics and Connect service metrics. The merged
+	// metrics server will be shut down when a signal is received by the main
+	// for loop using shutdownMetricsServer().
+	var server *http.Server
+	if c.flagEnableMetricsMerging {
+		c.logger.Info("Metrics is enabled, creating merged metrics server.")
+		server = c.createMergedMetricsServer()
+
+		// Run the merged metrics server
+		c.logger.Info("Running merged metrics server.")
+		go func() {
+			if err = server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
```

Comment:
If the metrics merging server fails to start (e.g. it can't bind to its port) should the sidecar container continue to run? I would have thought we'd error out so the user is aware.

Reply:
That's a good point. I suppose if a user misconfigures metrics with something like a conflicting port and there's only one log at the top of the consul-sidecar logs it will be fairly hidden to them that it's broken, so I think it makes sense to error out. I was reading [here](https://stackoverflow.com/questions/25920963/in-what-situation-will-http-listenandserve-return) that it seems common to `log.Fatal` here. `hclog` doesn't have a log.Fatal function but we can just `os.Exit(1)` after logging the error.

Reply:
This is exactly what happened to me. I mixed up the scrape port and merged port annotations and I set the merged port to be equal to the scrape port. Then the merged server failed to start because it tried to bind to a port that was already bound to (since that was the port my app was exposing its metrics on).